sdk/iot/core/src/az_iot_hub_client.c

```diff
@@ -107,3 +108,40 @@ AZ_NODISCARD az_result az_iot_hub_client_id_get(
 
   return AZ_OK;
 }
+
+AZ_NODISCARD az_result
+az_iot_hub_client_properties_init(az_iot_hub_client_properties* properties, az_span buffer)
+{
+  AZ_PRECONDITION_NOT_NULL(properties);
+  AZ_PRECONDITION_VALID_SPAN(buffer, 0, false);
+
+  properties->_internal.properties = buffer;
+  properties->_internal.current_property = az_span_ptr(buffer);
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_hub_client_properties_append(
+    az_iot_hub_client_properties* properties,
+    az_span name,
+    az_span value)
+{
+  AZ_PRECONDITION_NOT_NULL(properties);
+  AZ_PRECONDITION_VALID_SPAN(name, 1, false);
+  AZ_PRECONDITION_VALID_SPAN(value, 1, false);
+
+  az_span prop_span = properties->_internal.properties;
+
+  if (az_span_length(prop_span) > 0)
+  {
+    AZ_RETURN_IF_FAILED(az_span_append_uint8(prop_span, hub_client_param_separator, &prop_span));
+  }
+
+  AZ_RETURN_IF_FAILED(az_span_append(prop_span, name, &prop_span));
```

Comment:
Do we expected the `name` and `value` already RFC 2396-encoded as per https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#sending-device-to-cloud-messages?

Reply:
As per my current understanding yes. @ericwol-msft can confirm.

Reply:
Should convenience layer do the encoding or we reply on user application?
Should we check validity the encoding before sending telemetry to cloud?
What is the expected result from user application's perspective when the encoding is incorrect?

Reply:
1) I don't think that should be a baked in functionality to this API. That would mean someone that knows they have it encoded correctly would have to go through the same API and it would then do similar checks/validations unnecessarily.
2) We should not validate. Let the service do the grunt work of validating. Less cycles on the device and less code size.
3) Expect the service to fail the request/operation.

Reply:
When the service fails the request, how does user application know what is going on? Correct me if I was wrong. The service would disconnect from device at TCP layer if the property encoding is incorrect. There is no mechanism at MQTT layer to return error status on publish. So what user application would see is, the connection is disconnected. Should user application continue with a clean session or not (whether to retransmit telemetry data)?

Reply:
I think I stand to be corrected.
Based on the MQTT spec I don't think it's required to be RFC2396 encoded. Should be UTF-8 RFC3629.
[Source here](https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.pdf#%5B%7B%22num%22%3A127%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C105%2C546%2C0%5D)

To back this, I tried using MQTT Explorer and sent a message with properties that weren't URL encoded and everything seemed to work. Now I don't know if it is doing underlying manipulation but it does seem to work.

Reply:
You are correct. I double verified escape values in topic. Only "/" is not supported by service and service will disconnect immediately. But why the link I pasted above says the property should be url-encoded? Is it optional or mandatory?

Reply:
I don't know but thank you for bringing this up. I'll maybe reach out to someone from the service team to get "the truth".

Reply:
Regardless of what the ultimate decision is (encode or not), it won't be baked into this API. Will take a note but resolve for now.

Reply:
It was a design decision to not encode. If Unicode support is required we would need to use wchar_t, allow higher Uniocode code-points, etc.
This is still possible today as the application developer can bake them in the application. If they change dynamically then they will need to obtain a proper URI-encoder which is non-trivial and may be quite large.

BTW: we expect the same issues with the host-name part of the URI. @TiejunMS does NX support IDN such as these for DNS resolution today: https://www.w3.org/2003/Talks/0425-duerst-idniri/slide12-0.html?

Reply:
We don't encode domain name in DNS.
Add @hihigupt , in this case, I don't think we should encode properties in convenience layer either. What do you think?